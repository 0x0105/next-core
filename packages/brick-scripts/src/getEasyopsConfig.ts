import fs from "fs-extra"
import path from "path"

export const getEasyopsConfig = () => {
  let easyopsConfig = {
    contractYamlDir: "easyops",
    contractUrl: "git@git.easyops.local:anyclouds/contract-center.git",
    useLocalSdk: false
  }
  const isEasyopsConfigExists = fs.existsSync(path.join(process.cwd(), ".easyops-yo.json"))
  if (isEasyopsConfigExists) {
    const incomeConfig = fs.readJsonSync(path.join(process.cwd(), ".easyops-yo.json"))
    easyopsConfig = Object.assign(easyopsConfig, incomeConfig)
  }
  return easyopsConfig
}
