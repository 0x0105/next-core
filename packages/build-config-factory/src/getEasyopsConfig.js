const fs = require("fs-extra");
const path = require("path")

module.exports = {
  easyopsConfig: fs.readJsonSync(path.join(process.cwd(), ".easyops-yo.json")),
  isEasyopsConfigExists: fs.existsSync(path.join(process.cwd(), ".easyops-yo.json"))
}
