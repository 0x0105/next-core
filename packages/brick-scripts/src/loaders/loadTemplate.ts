import fs from "fs-extra";
import os from "os";
import path from "path";
import klawSync from "klaw-sync";
import * as changeCase from "change-case";
import rp from "request-promise-native";
import { FileWithContent, TargetType } from "../interface";
import { getEasyopsConfig } from "../getEasyopsConfig";
const { exec } = require("child_process");
// `tsc` will compile files which `import` or `require`,
// thus, we read file content instead of importing.
const easyopsConfig = getEasyopsConfig();
const packageJson = JSON.parse(
  fs.readFileSync(path.join(__dirname, "../../package.json"), "utf8")
);
const { templateDependencies } = packageJson;
const basicBricksVersion = templateDependencies["@bricks/basic-bricks"];
const brickContainerVersion = templateDependencies["@easyops/brick-container"];

const workspacePackageJson = JSON.parse(
  fs.readFileSync(path.join(process.cwd(), "package.json"), "utf8")
);
const workspaceHomepage = workspacePackageJson.homepage;
let userName = "bot";
exec("git config user.name", { encoding: "utf8" }, function (err, value) {
  if (err) return;
  userName = value.trim();
});

function escapeRegExp(str: string): string {
  return str.replace(/([.*+?^=!:${}()|[\]/\\])/g, "\\$1");
}

function replaceAll(str: string, find: string, replace: string): string {
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}

function replaceFileContent(
  filePath: string,
  translations: Record<string, string>
): string {
  const content = fs.readFileSync(filePath, "utf8");
  const contentFinds = Object.keys(translations).filter((key) =>
    content.includes(key)
  );
  if (contentFinds.length > 0) {
    return contentFinds.reduce(
      (acc, find) => replaceAll(acc, find, translations[find]),
      content
    );
  }
  return content;
}

function replaceDepsVersion(jsonString: string): string {
  const pkg = JSON.parse(jsonString);

  const peerDeps = pkg.peerDependencies || {};
  if (peerDeps["@bricks/basic-bricks"]) {
    peerDeps["@bricks/basic-bricks"] = basicBricksVersion;
  }
  return JSON.stringify(pkg, null, 2) + os.EOL;
}

export async function loadTemplate({
  targetType,
  packageName,
  brickName,
  templateName,
  processorName,
  targetRoot,
  docRoot,
}: {
  targetType: TargetType;
  packageName: string;
  brickName: string;
  templateName: string;
  processorName: string;
  targetRoot: string;
  docRoot: string;
}): Promise<FileWithContent[]> {
  const targetMap: { [key: string]: string } = {
    [TargetType.A_NEW_BRICK]: "brick",
    [TargetType.A_NEW_CUSTOM_TEMPLATE]: "custom-template",
    [TargetType.A_NEW_PACKAGE_OF_BRICKS]: "bricks-pkg",
    [TargetType.A_NEW_PACKAGE_OF_LIBS]: "libs-pkg",
    [TargetType.A_NEW_PACKAGE_OF_MICRO_APPS]: "micro-apps-pkg",
    [TargetType.A_NEW_CUSTOM_PROVIDER_BRICK]: "custom-provider-brick",
    [TargetType.A_NEW_CUSTOM_PROCESSOR]: "custom-processor",
    [TargetType.A_NEW_PACKAGE_OF_PROVIDERS]: "providers-pkg",
    [TargetType.A_NEW_PACKAGE_OF_DLL]: "dll-pkg",
    [TargetType.TRANSFORM_A_MICRO_APP]: "transformed-micro-apps-pkg",
    [TargetType.A_NEW_LEGACY_TEMPLATE]: "template",
    [TargetType.A_NEW_PACKAGE_OF_LEGACY_TEMPLATES]: "templates-pkg",
    [TargetType.I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES]:
      "i18n-patched-templates-pkg",
  };
  const templateDirOfFileName = targetMap[targetType];
  const templateRoot = path.join(__dirname, "../../template");
  let templateDir = path.join(templateRoot, templateDirOfFileName);

  const ignores = [".DS_Store"];
  let sdkName: string;
  if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库直接使用构件库的模板文件，但忽略其 `src` 目录。
    templateDir = path.join(
      templateRoot,
      targetMap[TargetType.A_NEW_PACKAGE_OF_BRICKS]
    );
    ignores.push("template/bricks-pkg/src/");
    sdkName = `${packageName.replace(/^providers-of-/, "")}-sdk`;
  }

  const translations: Record<string, string> = {
    "$workspace-homepage$": workspaceHomepage,
    "$kebab-package-name$": packageName,
    $PascalPackageName$: changeCase.pascalCase(packageName),
    "$Title Package Name$": changeCase.capitalCase(packageName),
    $CONSTANT_PACKAGE_NAME$: changeCase.constantCase(packageName),
    $camelCasePackageName$: changeCase.camelCase(packageName),
    $PascalBrickName$: changeCase.pascalCase(brickName),
    "$kebab-brick-name$": `${packageName}.${brickName}`,
    "$kebab-brick-last-name$": brickName,
    "$kebab-custom-provider-brick-name$": `${packageName}.provider-${brickName}`,
    "$generator.version$": `v${packageJson.version}`,
    "$brick.container.version$": brickContainerVersion,
    "$kebab-sdk-name$": sdkName,
    "$kebab-template-name$": templateName,
    "$kebab-username$": userName,
    $camelTemplateName$: changeCase.camelCase(templateName),
    $PascalTemplateName$: changeCase.pascalCase(templateName),
    $camelProcessorName$: changeCase.camelCase(processorName),
  };

  const filter = (src: string): boolean =>
    ignores.every((item) => !src.includes(item));

  const templateGroups: any = [
    {
      templateDir,
      targetDir: targetRoot,
      files: klawSync(templateDir, {
        depthLimit: 4,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    },
  ];

  if (targetType === TargetType.A_NEW_PACKAGE_OF_BRICKS && brickName) {
    // Also create a new brick for the new bricks-package
    const brickTemplateDir = path.join(templateRoot, "brick");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: path.join(targetRoot, "src", brickName),
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });
  } else if (targetType === TargetType.A_NEW_PACKAGE_OF_LEGACY_TEMPLATES) {
    // Also create a new brick for the new bricks-package
    const templateTemplateDir = path.join(templateRoot, "template");
    templateGroups.push({
      templateDir: templateTemplateDir,
      targetDir: path.join(targetRoot, "src"),
      files: klawSync(templateTemplateDir, {
        depthLimit: 1,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });
  } else if (targetType === TargetType.A_NEW_PACKAGE_OF_PROVIDERS) {
    // Providers 库还有专属的一些模板文件。
    const brickTemplateDir = path.join(templateRoot, "providers-pkg");
    templateGroups.push({
      templateDir: brickTemplateDir,
      targetDir: targetRoot,
      files: klawSync(brickTemplateDir, {
        depthLimit: 2,
        nodir: true,
        filter: (item) => filter(item.path),
      }),
    });

    let sdkVersion;
    const getVersion = async () => {
      try {
        const sdkPackage = await rp({
          url: `https://registry.npm.easyops.local/@sdk/${sdkName}`,
          json: true,
          strictSSL: false,
        });
        sdkVersion = `^${sdkPackage["dist-tags"].latest}`;
      } catch {
        sdkVersion = "FETCH LATEST VERSION ERROR";
      }
      translations["$sdk.version$"] = sdkVersion;
    };

    if (!easyopsConfig?.useLocalSdk) {
      await getVersion();
    } else {
      translations["$sdk.version$"] = "1.0.0";
    }
  }

  const files: FileWithContent[] = templateGroups.reduce(
    (acc: FileWithContent[], group: any) =>
      acc.concat(
        group.files.map((file: any) => {
          const content = replaceFileContent(file.path, translations);

          let realPath = file.path;
          const basename = path.basename(file.path);
          const filenameFinds = Object.keys(translations).filter((key) =>
            basename.includes(key)
          );
          if (filenameFinds.length > 0) {
            realPath = path.join(
              path.dirname(realPath),
              filenameFinds.reduce(
                (acc, find) => replaceAll(acc, find, translations[find]),
                basename
              )
            );
          }

          return [
            path.join(
              group.targetDir,
              path.relative(group.templateDir, realPath)
            ),
            content,
          ];
        })
      ),
    []
  );

  if (
    targetType !== TargetType.A_NEW_BRICK &&
    targetType !== TargetType.A_NEW_CUSTOM_TEMPLATE &&
    targetType !== TargetType.A_NEW_CUSTOM_PROVIDER_BRICK &&
    targetType !== TargetType.A_NEW_CUSTOM_PROCESSOR &&
    targetType !== TargetType.A_NEW_LEGACY_TEMPLATE &&
    targetType !== TargetType.TRANSFORM_A_MICRO_APP &&
    targetType !== TargetType.I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES
  ) {
    files.push([
      path.join(targetRoot, "package.json"),
      replaceDepsVersion(
        replaceFileContent(
          path.join(templateRoot, `${targetMap[targetType]}.json`),
          translations
        )
      ),
    ]);
  }

  return files;
}
