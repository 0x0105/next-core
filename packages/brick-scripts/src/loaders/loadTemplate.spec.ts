import path from "path";
import { loadTemplate } from "./loadTemplate";
import { FileWithContent, TargetType } from "../interface";
import { getEasyopsConfig } from "../getEasyopsConfig";

jest.mock("../getEasyopsConfig");
(getEasyopsConfig as jest.Mock).mockReturnValue({
  contractYamlDir: "easyops",
  contractUrl: "https://github.com/easyops-cn/contract-center.git",
  useLocalSdk: true,
});

jest.mock("request-promise-native");

function ignoreVersionRelatedFiles(
  files: FileWithContent[]
): FileWithContent[] {
  return files.filter(([filePath]) => {
    const fileName = path.basename(filePath);
    return !["package.json", "README.md", "package.conf.yaml"].includes(
      fileName
    );
  });
}

describe("loadTemplate", () => {
  it("should create a new brick", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_BRICK,
      packageName: "for-good",
      brickName: "for-better",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new editor brick", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_EDITOR_BRICK,
      packageName: "for-good",
      brickName: "for-better",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new custom template", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_CUSTOM_TEMPLATE,
      packageName: "for-good",
      brickName: "for-better",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new package of bricks", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_BRICKS,
      packageName: "for-good",
      brickName: "for-better",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new package of bricks without any bricks", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_BRICKS,
      packageName: "for-good",
      brickName: "",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new package of libs", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_LIBS,
      packageName: "for-good",
      brickName: "",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new package of micro-apps", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_MICRO_APPS,
      packageName: "for-good",
      brickName: "",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new custom provider brick", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_CUSTOM_PROVIDER_BRICK,
      packageName: "for-good",
      brickName: "get-data",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new custom processor", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_CUSTOM_PROCESSOR,
      packageName: "for-good",
      brickName: "",
      templateName: "",
      processorName: "doGood",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new package of providers", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_PROVIDERS,
      packageName: "for-good",
      brickName: "",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new package of dll", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_DLL,
      packageName: "for-good",
      brickName: "",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should transform a micro-app", async () => {
    const files = await loadTemplate({
      targetType: TargetType.TRANSFORM_A_MICRO_APP,
      packageName: "for-good",
      brickName: "",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should create a new legacy template", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_LEGACY_TEMPLATE,
      packageName: "for-good",
      brickName: "",
      templateName: "for-better",
      processorName: "",
      targetRoot: "dist",
    });
    expect(files).toMatchSnapshot();
  });
  it("should create a new package of legacy templates", async () => {
    const files = await loadTemplate({
      targetType: TargetType.A_NEW_PACKAGE_OF_LEGACY_TEMPLATES,
      packageName: "for-good",
      brickName: "",
      templateName: "for-better",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
  it("should i18n-patch a package of legacy templates", async () => {
    const files = await loadTemplate({
      targetType: TargetType.I18N_PATCH_A_PACKAGE_OF_LEGACY_TEMPLATES,
      packageName: "for-good",
      brickName: "",
      templateName: "",
      processorName: "",
      targetRoot: "dist",
    });
    const otherFiles = ignoreVersionRelatedFiles(files);
    expect(otherFiles).toMatchSnapshot();
  });
});
