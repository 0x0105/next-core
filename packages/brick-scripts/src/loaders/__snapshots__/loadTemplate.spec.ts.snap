// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadTemplate should create a new brick 1`] = `
Array [
  Array [
    "dist/ForBetter.spec.tsx",
    "import React from \\"react\\";
import { shallow } from \\"enzyme\\";
import { ForBetter } from \\"./ForBetter\\";

describe(\\"ForBetter\\", () => {
  it(\\"should work\\", () => {
    const wrapper = shallow(<ForBetter />);
    expect(wrapper).toMatchSnapshot();
  });
});
",
  ],
  Array [
    "dist/ForBetter.tsx",
    "import React from \\"react\\";
import { useTranslation } from \\"react-i18next\\";
import { NS_FOR_GOOD, K } from \\"../i18n/constants\\";

export function ForBetter(): React.ReactElement {
  const { t } = useTranslation(NS_FOR_GOOD);

  return <div>{t(K.FOR_GOOD)} works!</div>;
}
",
  ],
  Array [
    "dist/__snapshots__/ForBetter.spec.tsx.snap",
    "// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[\`ForBetter should work 1\`] = \`
<div>
  FOR_GOOD
   works!
</div>
\`;
",
  ],
  Array [
    "dist/index.spec.ts",
    "// Ref https://github.com/jsdom/jsdom/issues/1030
import \\"document-register-element\\";
import ReactDOM from \\"react-dom\\";
import \\"./\\";

const spyOnRender = jest.spyOn(ReactDOM, \\"render\\").mockImplementation(() => {});
const unmountComponentAtNode = jest
  .spyOn(ReactDOM, \\"unmountComponentAtNode\\")
  .mockImplementation((() => {}) as any);

describe(\\"for-good.for-better\\", () => {
  it(\\"should create a custom element\\", async () => {
    const element = document.createElement(\\"for-good.for-better\\");
    // Always waiting for async \`(dis)connectedCallback\`
    await jest.runAllTimers();
    expect(spyOnRender).not.toBeCalled();
    document.body.appendChild(element);
    await jest.runAllTimers();
    expect(spyOnRender).toBeCalled();
    document.body.removeChild(element);
    await jest.runAllTimers();
    expect(unmountComponentAtNode).toBeCalled();
  });
});
",
  ],
  Array [
    "dist/index.tsx",
    "import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { BrickWrapper } from \\"@easyops/brick-kit\\";
import { ForBetter } from \\"./ForBetter\\";

class ForBetterElement extends HTMLElement {
  connectedCallback(): void {
    this.style.display = \\"block\\";
    this._render();
  }

  disconnectedCallback(): void {
    ReactDOM.unmountComponentAtNode(this);
  }

  private _render(): void {
    if (this.isConnected) {
      ReactDOM.render(
        <BrickWrapper>
          <ForBetter />
        </BrickWrapper>,
        this
      );
    }
  }
}

customElements.define(\\"for-good.for-better\\", ForBetterElement);
",
  ],
  Array [
    "doc/for-good/for-better.md",
    "# INPUTS

| property | type | required | default | description |
| -------- | ---- | -------- | ------- | ----------- |
| -        | -    | -        | -       | -           |

# EVENTS

| type | detail | description |
| ---- | ------ | ----------- |
| -    | -      | -           |

# METHODS

| name | params | description |
| ---- | ------ | ----------- |
| -    | -      | -           |
",
  ],
]
`;

exports[`loadTemplate should create a new package of bricks 1`] = `
Array [
  Array [
    "dist/scripts/post-build.js",
    "const { postBuild } = require(\\"@easyops/build-config-factory\\");

postBuild(\\"bricks\\");
",
  ],
  Array [
    "dist/src/i18n/constants.ts",
    "export const NS_FOR_GOOD = \\"for-good\\";

export enum K {
  FOR_GOOD = \\"FOR_GOOD\\"
}

export type Locale = { [key in K]: string };
",
  ],
  Array [
    "dist/src/i18n/index.ts",
    "import i18next from \\"i18next\\";
import { NS_FOR_GOOD } from \\"./constants\\";
import en from \\"./locales/en\\";
import zh from \\"./locales/zh\\";

i18next.addResourceBundle(\\"en\\", NS_FOR_GOOD, en);
i18next.addResourceBundle(\\"zh\\", NS_FOR_GOOD, zh);
",
  ],
  Array [
    "dist/src/i18n/locales/en.ts",
    "import { Locale, K } from \\"../constants\\";

const locale: Locale = {
  [K.FOR_GOOD]: \\"For Good\\"
};

export default locale;
",
  ],
  Array [
    "dist/src/i18n/locales/zh.ts",
    "import { Locale, K } from \\"../constants\\";

const locale: Locale = {
  [K.FOR_GOOD]: \\"For Good\\"
};

export default locale;
",
  ],
  Array [
    "dist/src/index.spec.ts",
    "import i18next from \\"i18next\\";
const spyOnAddResourceBundle = (i18next.addResourceBundle = jest.fn());

const spyOnDefine = jest.fn();
(window as any).customElements = {
  define: spyOnDefine
};

// Use \`require\` instead of \`import\` to avoid hoisting.
require(\\"./index\\");

describe(\\"index\\", () => {
  it(\\"should add i18n resource bundle\\", () => {
    expect(spyOnAddResourceBundle).toBeCalled();
  });
  it(\\"should define custom elements\\", () => {
    expect(spyOnDefine).toBeCalled();
  });
});
",
  ],
  Array [
    "dist/src/index.ts",
    "import \\"./i18n\\";
",
  ],
  Array [
    "dist/tsconfig.json",
    "{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"include\\": [
    \\"./src\\",
    \\"../../declarations\\"
  ]
}
",
  ],
  Array [
    "dist/webpack.config.js",
    "const merge = require(\\"webpack-merge\\");
const { bricks } = require(\\"@easyops/webpack-config-factory\\");

const { webpackCommonFactory, webpackDevFactory, webpackProdFactory } = bricks;

module.exports = merge(
  webpackCommonFactory(),
  process.env.NODE_ENV === \\"development\\"
    ? webpackDevFactory()
    : webpackProdFactory()
);
",
  ],
  Array [
    "dist/src/for-better/ForBetter.spec.tsx",
    "import React from \\"react\\";
import { shallow } from \\"enzyme\\";
import { ForBetter } from \\"./ForBetter\\";

describe(\\"ForBetter\\", () => {
  it(\\"should work\\", () => {
    const wrapper = shallow(<ForBetter />);
    expect(wrapper).toMatchSnapshot();
  });
});
",
  ],
  Array [
    "dist/src/for-better/ForBetter.tsx",
    "import React from \\"react\\";
import { useTranslation } from \\"react-i18next\\";
import { NS_FOR_GOOD, K } from \\"../i18n/constants\\";

export function ForBetter(): React.ReactElement {
  const { t } = useTranslation(NS_FOR_GOOD);

  return <div>{t(K.FOR_GOOD)} works!</div>;
}
",
  ],
  Array [
    "dist/src/for-better/__snapshots__/ForBetter.spec.tsx.snap",
    "// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[\`ForBetter should work 1\`] = \`
<div>
  FOR_GOOD
   works!
</div>
\`;
",
  ],
  Array [
    "dist/src/for-better/index.spec.ts",
    "// Ref https://github.com/jsdom/jsdom/issues/1030
import \\"document-register-element\\";
import ReactDOM from \\"react-dom\\";
import \\"./\\";

const spyOnRender = jest.spyOn(ReactDOM, \\"render\\").mockImplementation(() => {});
const unmountComponentAtNode = jest
  .spyOn(ReactDOM, \\"unmountComponentAtNode\\")
  .mockImplementation((() => {}) as any);

describe(\\"for-good.for-better\\", () => {
  it(\\"should create a custom element\\", async () => {
    const element = document.createElement(\\"for-good.for-better\\");
    // Always waiting for async \`(dis)connectedCallback\`
    await jest.runAllTimers();
    expect(spyOnRender).not.toBeCalled();
    document.body.appendChild(element);
    await jest.runAllTimers();
    expect(spyOnRender).toBeCalled();
    document.body.removeChild(element);
    await jest.runAllTimers();
    expect(unmountComponentAtNode).toBeCalled();
  });
});
",
  ],
  Array [
    "dist/src/for-better/index.tsx",
    "import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
import { BrickWrapper } from \\"@easyops/brick-kit\\";
import { ForBetter } from \\"./ForBetter\\";

class ForBetterElement extends HTMLElement {
  connectedCallback(): void {
    this.style.display = \\"block\\";
    this._render();
  }

  disconnectedCallback(): void {
    ReactDOM.unmountComponentAtNode(this);
  }

  private _render(): void {
    if (this.isConnected) {
      ReactDOM.render(
        <BrickWrapper>
          <ForBetter />
        </BrickWrapper>,
        this
      );
    }
  }
}

customElements.define(\\"for-good.for-better\\", ForBetterElement);
",
  ],
  Array [
    "doc/for-good/for-better.md",
    "# INPUTS

| property | type | required | default | description |
| -------- | ---- | -------- | ------- | ----------- |
| -        | -    | -        | -       | -           |

# EVENTS

| type | detail | description |
| ---- | ------ | ----------- |
| -    | -      | -           |

# METHODS

| name | params | description |
| ---- | ------ | ----------- |
| -    | -      | -           |
",
  ],
]
`;

exports[`loadTemplate should create a new package of dll 1`] = `
Array [
  Array [
    "dist/webpack.dll.js",
    "const { dll } = require(\\"@easyops/webpack-config-factory\\");

module.exports = dll.webpackDllFactory();
",
  ],
]
`;

exports[`loadTemplate should create a new package of libs 1`] = `
Array [
  Array [
    "dist/rollup.config.js",
    "import { rollupFactory } from \\"@easyops/rollup-config-factory\\";

export default rollupFactory({ umdName: \\"ForGood\\" });
",
  ],
  Array [
    "dist/src/index.ts",
    "",
  ],
  Array [
    "dist/tsconfig.json",
    "{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"include\\": [
    \\"./src\\",
    \\"../../declarations\\"
  ]
}
",
  ],
]
`;

exports[`loadTemplate should create a new package of micro-apps 1`] = `
Array [
  Array [
    "dist/scripts/post-build.js",
    "const { postBuild } = require(\\"@easyops/build-config-factory\\");

postBuild(\\"micro-apps\\");
",
  ],
  Array [
    "dist/storyboard.json",
    "{
  \\"$schema\\": \\"../../node_modules/@easyops/brick-types/.schema/storyboard.json\\",
  \\"imports\\": [\\"@bricks/basic-bricks\\"],
  \\"app\\": {
    \\"name\\": \\"For Good\\",
    \\"id\\": \\"for-good\\",
    \\"homepage\\": \\"/for-good\\"
  },
  \\"routes\\": []
}
",
  ],
]
`;

exports[`loadTemplate should create a new package of providers 1`] = `
Array [
  Array [
    "dist/scripts/post-build.js",
    "const { postBuild } = require(\\"@easyops/build-config-factory\\");

postBuild(\\"bricks\\");
",
  ],
  Array [
    "dist/tsconfig.json",
    "{
  \\"extends\\": \\"../../tsconfig.json\\",
  \\"include\\": [
    \\"./src\\",
    \\"../../declarations\\"
  ]
}
",
  ],
  Array [
    "dist/webpack.config.js",
    "const merge = require(\\"webpack-merge\\");
const { bricks } = require(\\"@easyops/webpack-config-factory\\");

const { webpackCommonFactory, webpackDevFactory, webpackProdFactory } = bricks;

module.exports = merge(
  webpackCommonFactory(),
  process.env.NODE_ENV === \\"development\\"
    ? webpackDevFactory()
    : webpackProdFactory()
);
",
  ],
  Array [
    "dist/.gitignore",
    "/src/
",
  ],
  Array [
    "dist/scripts/pre-build.js",
    "const { preBuild } = require(\\"@easyops/build-config-factory\\");

preBuild(\\"providers\\");
",
  ],
]
`;

exports[`loadTemplate should create a new package of templates 1`] = `
Array [
  Array [
    "dist/deploy-default/pacakge.conf.yaml",
    "install_path: /usr/local/easyops/templates/for-good-NT
user: easyops:easyops
dependencies:
  - name: brick_next
    version: \\">=0.19.1\\"
    local_deploy: true
",
  ],
  Array [
    "dist/scripts/post-build.js",
    "const { postBuild } = require(\\"@easyops/build-config-factory\\");

postBuild(\\"templates\\");
",
  ],
  Array [
    "dist/webpack.config.js",
    "const merge = require(\\"webpack-merge\\");
const { bricks } = require(\\"@easyops/webpack-config-factory\\");

const { webpackCommonFactory, webpackDevFactory, webpackProdFactory } = bricks;

module.exports = merge(
  webpackCommonFactory({
    scope: \\"templates\\"
  }),
  process.env.NODE_ENV === \\"development\\"
    ? webpackDevFactory()
    : webpackProdFactory()
);
",
  ],
  Array [
    "dist/src/for-better.spec.ts",
    "import { BrickConf } from \\"@easyops/brick-types\\";
import {
  $camelTemplateNameFactory,
  $PascalTemplateParams
} from \\"./for-better\\";

jest.mock(\\"@easyops/brick-kit\\", () => ({
  getRuntime: () => ({
    registerBrickTemplate: jest.fn()
  })
}));

describe(\\"micro-app\\", () => {
  it.each<[$PascalTemplateParams, BrickConf]>([
    [
      {},
      {
        brick: \\"your.brick\\"
      }
    ]
  ])(\\"$camelTemplateNameFactory(%j) should work\\", (params, brickConf) => {
    expect($camelTemplateNameFactory(params)).toEqual(brickConf);
  });
});
",
  ],
  Array [
    "dist/src/for-better.ts",
    "import { getRuntime } from \\"@easyops/brick-kit\\";
import { BrickConf } from \\"@easyops/brick-types\\";

export interface $PascalTemplateParams {}

export function $camelTemplateNameFactory(
  params: $PascalTemplateParams
): BrickConf {
  return {
    brick: \\"your.brick\\"
  };
}

getRuntime().registerBrickTemplate(
  \\"for-good.for-better\\",
  $camelTemplateNameFactory
);
",
  ],
]
`;

exports[`loadTemplate should create a new template 1`] = `
Array [
  Array [
    "dist/for-better.spec.ts",
    "import { BrickConf } from \\"@easyops/brick-types\\";
import {
  $camelTemplateNameFactory,
  $PascalTemplateParams
} from \\"./for-better\\";

jest.mock(\\"@easyops/brick-kit\\", () => ({
  getRuntime: () => ({
    registerBrickTemplate: jest.fn()
  })
}));

describe(\\"micro-app\\", () => {
  it.each<[$PascalTemplateParams, BrickConf]>([
    [
      {},
      {
        brick: \\"your.brick\\"
      }
    ]
  ])(\\"$camelTemplateNameFactory(%j) should work\\", (params, brickConf) => {
    expect($camelTemplateNameFactory(params)).toEqual(brickConf);
  });
});
",
  ],
  Array [
    "dist/for-better.ts",
    "import { getRuntime } from \\"@easyops/brick-kit\\";
import { BrickConf } from \\"@easyops/brick-types\\";

export interface $PascalTemplateParams {}

export function $camelTemplateNameFactory(
  params: $PascalTemplateParams
): BrickConf {
  return {
    brick: \\"your.brick\\"
  };
}

getRuntime().registerBrickTemplate(
  \\"for-good.for-better\\",
  $camelTemplateNameFactory
);
",
  ],
]
`;

exports[`loadTemplate should transform a micro-app 1`] = `
Array [
  Array [
    "dist/.gitignore",
    "/storyboard.json
",
  ],
  Array [
    "dist/scripts/build.js",
    "const { build } = require(\\"@easyops/build-config-factory\\");

build(\\"micro-apps\\");
",
  ],
  Array [
    "dist/tsconfig.json",
    "{
  \\"compilerOptions\\": {
    \\"target\\": \\"ES6\\",
    \\"module\\": \\"CommonJS\\",
    \\"moduleResolution\\": \\"node\\",
    \\"esModuleInterop\\": true,
    \\"resolveJsonModule\\": true,
    \\"skipLibCheck\\": true,
    \\"declaration\\": false,
    \\"outDir\\": \\"./dist\\"
  },
  \\"exclude\\": [
    \\"node_modules\\",
    \\"**/*.spec.ts\\",
    \\"dist/\\",
    \\"template/\\",
    \\"**/__mocks__/\\",
    \\"package.json\\"
  ]
}
",
  ],
]
`;
